File: src/rust/rust-proto/src/graplinc/grapl/api/plugin_sdk/analyzers/v1beta1/messages.rs
use std::time::SystemTime;

use crate::{
    graplinc::grapl::{
        api::graph_query_service::v1beta1::messages::GraphView,
        common::v1beta1::types::{
            EdgeName,
            PropertyName,
            Uid,
        },
    },
    protobufs::graplinc::grapl::api::plugin_sdk::analyzers::v1beta1::{
        execution_result::Inner as ExecutionResultInnerProto,
        update::Inner as UpdateInnerProto,
        AnalyzerName as AnalyzerNameProto,
        EdgeUpdate as EdgeUpdateProto,
        ExecutionHit as ExecutionHitProto,
        ExecutionMiss as ExecutionMissProto,
        ExecutionResult as ExecutionResultProto,
        Int64PropertyUpdate as Int64PropertyUpdateProto,
        LensRef as LensRefProto,
        RunAnalyzerRequest as RunAnalyzerRequestProto,
        RunAnalyzerResponse as RunAnalyzerResponseProto,
        StringPropertyUpdate as StringPropertyUpdateProto,
        UInt64PropertyUpdate as UInt64PropertyUpdateProto,
        Update as UpdateProto,
        Updates as UpdatesProto,
    },
    serde_impl::ProtobufSerializable,
    type_url,
    SerDeError,
};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct StringPropertyUpdate {
    pub uid: Uid,
    pub property_name: PropertyName,
}

impl TryFrom<StringPropertyUpdateProto> for StringPropertyUpdate {
    type Error = SerDeError;
    fn try_from(value: StringPropertyUpdateProto) -> Result<Self, Self::Error> {
        Ok(Self {
            uid: value
                .uid
                .ok_or(SerDeError::MissingField("uid"))?
                .try_into()?,
            property_name: value
                .property_name
                .ok_or(SerDeError::MissingField("property_name"))?
                .try_into()?,
        })
    }
}

impl From<StringPropertyUpdate> for StringPropertyUpdateProto {
    fn from(value: StringPropertyUpdate) -> Self {
        Self {
            uid: Some(value.uid.into()),
            property_name: Some(value.property_name.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct UInt64PropertyUpdate {
    pub uid: Uid,
    pub property_name: PropertyName,
}

impl TryFrom<UInt64PropertyUpdateProto> for UInt64PropertyUpdate {
    type Error = SerDeError;
    fn try_from(value: UInt64PropertyUpdateProto) -> Result<Self, Self::Error> {
        Ok(Self {
            uid: value
                .uid
                .ok_or(SerDeError::MissingField("uid"))?
                .try_into()?,
            property_name: value
                .property_name
                .ok_or(SerDeError::MissingField("property_name"))?
                .try_into()?,
        })
    }
}

impl From<UInt64PropertyUpdate> for UInt64PropertyUpdateProto {
    fn from(value: UInt64PropertyUpdate) -> Self {
        Self {
            uid: Some(value.uid.into()),
            property_name: Some(value.property_name.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Int64PropertyUpdate {
    pub uid: Uid,
    pub property_name: PropertyName,
}

impl TryFrom<Int64PropertyUpdateProto> for Int64PropertyUpdate {
    type Error = SerDeError;
    fn try_from(value: Int64PropertyUpdateProto) -> Result<Self, Self::Error> {
        Ok(Self {
            uid: value
                .uid
                .ok_or(SerDeError::MissingField("uid"))?
                .try_into()?,
            property_name: value
                .property_name
                .ok_or(SerDeError::MissingField("property_name"))?
                .try_into()?,
        })
    }
}

impl From<Int64PropertyUpdate> for Int64PropertyUpdateProto {
    fn from(value: Int64PropertyUpdate) -> Self {
        Self {
            uid: Some(value.uid.into()),
            property_name: Some(value.property_name.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct EdgeUpdate {
    pub src_uid: Uid,
    pub dst_uid: Uid,
    pub forward_edge_name: EdgeName,
    pub reverse_edge_name: EdgeName,
}

impl TryFrom<EdgeUpdateProto> for EdgeUpdate {
    type Error = SerDeError;
    fn try_from(value: EdgeUpdateProto) -> Result<Self, Self::Error> {
        Ok(Self {
            src_uid: value
                .src_uid
                .ok_or(SerDeError::MissingField("EdgeUpdate.forward_edge_name"))?
                .try_into()?,
            dst_uid: value
                .dst_uid
                .ok_or(SerDeError::MissingField("EdgeUpdate.reverse_edge_name"))?
                .try_into()?,
            forward_edge_name: value
                .forward_edge_name
                .ok_or(SerDeError::MissingField("EdgeUpdate.forward_edge_name"))?
                .try_into()?,
            reverse_edge_name: value
                .reverse_edge_name
                .ok_or(SerDeError::MissingField("EdgeUpdate.reverse_edge_name"))?
                .try_into()?,
        })
    }
}

impl From<EdgeUpdate> for EdgeUpdateProto {
    fn from(value: EdgeUpdate) -> Self {
        Self {
            src_uid: Some(value.src_uid.into()),
            dst_uid: Some(value.dst_uid.into()),
            forward_edge_name: Some(value.forward_edge_name.into()),
            reverse_edge_name: Some(value.reverse_edge_name.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Update {
    StringProperty(StringPropertyUpdate),
    Uint64Property(UInt64PropertyUpdate),
    Int64Property(Int64PropertyUpdate),
    Edge(EdgeUpdate),
}

impl TryFrom<UpdateProto> for Update {
    type Error = SerDeError;
    fn try_from(value: UpdateProto) -> Result<Self, Self::Error> {
        match value.inner {
            Some(UpdateInnerProto::StringProperty(update)) => {
                Ok(Update::StringProperty(update.try_into()?))
            }
            Some(UpdateInnerProto::Uint64Property(update)) => {
                Ok(Update::Uint64Property(update.try_into()?))
            }
            Some(UpdateInnerProto::Int64Property(update)) => {
                Ok(Update::Int64Property(update.try_into()?))
            }
            Some(UpdateInnerProto::Edge(update)) => Ok(Update::Edge(update.try_into()?)),
            None => Err(SerDeError::UnknownVariant("Update")),
        }
    }
}

impl From<Update> for UpdateProto {
    fn from(value: Update) -> Self {
        match value {
            Update::StringProperty(update) => UpdateProto {
                inner: Some(UpdateInnerProto::StringProperty(update.into())),
            },
            Update::Uint64Property(update) => UpdateProto {
                inner: Some(UpdateInnerProto::Uint64Property(update.into())),
            },
            Update::Int64Property(update) => UpdateProto {
                inner: Some(UpdateInnerProto::Int64Property(update.into())),
            },
            Update::Edge(update) => UpdateProto {
                inner: Some(UpdateInnerProto::Edge(update.into())),
            },
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
pub struct Updates {
    pub updates: Vec<Update>,
}

impl Updates {
    pub fn new() -> Self {
        Self { updates: vec![] }
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            updates: Vec::with_capacity(capacity),
        }
    }

    pub fn iter(&self) -> std::slice::Iter<Update> {
        self.updates.iter()
    }
}

impl IntoIterator for Updates {
    type Item = Update;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.updates.into_iter()
    }
}

impl TryFrom<UpdatesProto> for Updates {
    type Error = SerDeError;
    fn try_from(value: UpdatesProto) -> Result<Self, Self::Error> {
        Ok(Self {
            updates: value
                .updates
                .into_iter()
                .map(|update| update.try_into())
                .collect::<Result<Vec<_>, _>>()?,
        })
    }
}

impl From<Updates> for UpdatesProto {
    fn from(value: Updates) -> Self {
        Self {
            updates: value
                .updates
                .into_iter()
                .map(|update| update.into())
                .collect(),
        }
    }
}

impl type_url::TypeUrl for Updates {
    const TYPE_URL: &'static str =
        "graplsecurity.com/graplinc.grapl.api.plugin_sdk.analyzers.v1beta1.Updates";
}

impl ProtobufSerializable for Updates {
    type ProtobufMessage = UpdatesProto;
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct LensRef {
    pub lens_namespace: String,
    pub lens_name: String,
}

impl TryFrom<LensRefProto> for LensRef {
    type Error = SerDeError;
    fn try_from(value: LensRefProto) -> Result<Self, Self::Error> {
        // todo: ensure lens_namespace and lens_name conform
        Ok(LensRef {
            lens_namespace: value.lens_namespace,
            lens_name: value.lens_name,
        })
    }
}

impl From<LensRef> for LensRefProto {
    fn from(value: LensRef) -> Self {
        LensRefProto {
            lens_namespace: value.lens_namespace,
            lens_name: value.lens_name,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct AnalyzerName {
    pub value: String,
}

impl TryFrom<AnalyzerNameProto> for AnalyzerName {
    type Error = SerDeError;
    fn try_from(value: AnalyzerNameProto) -> Result<Self, Self::Error> {
        // todo: Add check for length/ conformance defined in proto
        Ok(AnalyzerName { value: value.value })
    }
}

impl From<AnalyzerName> for AnalyzerNameProto {
    fn from(value: AnalyzerName) -> Self {
        AnalyzerNameProto { value: value.value }
    }
}

#[derive(Debug, Clone)]
pub struct ExecutionHit {
    pub graph_view: GraphView,
    pub lens_refs: Vec<LensRef>,
    pub analyzer_name: AnalyzerName,
    pub time_of_match: SystemTime,
    pub idempotency_key: u64,
    pub score: i32,
}

impl TryFrom<ExecutionHitProto> for ExecutionHit {
    type Error = SerDeError;
    fn try_from(value: ExecutionHitProto) -> Result<Self, Self::Error> {
        // todo: Add check for length/ conformance defined in proto
        Ok(Self {
            graph_view: value
                .graph_view
                .ok_or(SerDeError::MissingField("ExecutionHit.graph_view"))?
                .try_into()?,
            lens_refs: value
                .lens_refs
                .into_iter()
                .map(|lens_ref| lens_ref.try_into())
                .collect::<Result<Vec<_>, _>>()?,
            analyzer_name: value
                .analyzer_name
                .ok_or(SerDeError::MissingField("ExecutionHit.analyzer_name"))?
                .try_into()?,
            time_of_match: value
                .time_of_match
                .ok_or(SerDeError::MissingField("ExecutionHit.time_of_match"))?
                .try_into()?,
            idempotency_key: value.idempotency_key,
            score: value.score,
        })
    }
}

impl From<ExecutionHit> for ExecutionHitProto {
    fn from(value: ExecutionHit) -> Self {
        Self {
            graph_view: Some(value.graph_view.into()),
            lens_refs: value
                .lens_refs
                .into_iter()
                .map(|lens_ref| lens_ref.into())
                .collect(),
            analyzer_name: Some(value.analyzer_name.into()),
            time_of_match: Some(value.time_of_match.try_into().unwrap()), // this can never actually fail
            idempotency_key: value.idempotency_key.into(),
            score: value.score.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExecutionMiss {}

impl TryFrom<ExecutionMissProto> for ExecutionMiss {
    type Error = SerDeError;
    fn try_from(value: ExecutionMissProto) -> Result<Self, Self::Error> {
        let ExecutionMissProto {} = value;
        Ok(Self {})
    }
}

impl From<ExecutionMiss> for ExecutionMissProto {
    fn from(value: ExecutionMiss) -> Self {
        let ExecutionMiss {} = value;
        Self {}
    }
}

#[derive(Debug, Clone)]
pub enum ExecutionResult {
    ExecutionHit(ExecutionHit),
    ExecutionMiss(ExecutionMiss),
}

impl TryFrom<ExecutionResultProto> for ExecutionResult {
    type Error = SerDeError;
    fn try_from(value: ExecutionResultProto) -> Result<Self, Self::Error> {
        match value.inner {
            Some(ExecutionResultInnerProto::Hit(inner)) => {
                Ok(Self::ExecutionHit(inner.try_into()?))
            }
            Some(ExecutionResultInnerProto::Miss(inner)) => {
                Ok(Self::ExecutionMiss(inner.try_into()?))
            }
            None => Err(SerDeError::UnknownVariant("ExecutionResult")),
        }
    }
}

impl From<ExecutionResult> for ExecutionResultProto {
    fn from(value: ExecutionResult) -> Self {
        match value {
            ExecutionResult::ExecutionHit(value) => Self {
                inner: Some(ExecutionResultInnerProto::Hit(value.into())),
            },
            ExecutionResult::ExecutionMiss(value) => Self {
                inner: Some(ExecutionResultInnerProto::Miss(value.into())),
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RunAnalyzerRequest {
    pub tenant_id: uuid::Uuid,
    pub update: Update,
}

impl TryFrom<RunAnalyzerRequestProto> for RunAnalyzerRequest {
    type Error = SerDeError;
    fn try_from(value: RunAnalyzerRequestProto) -> Result<Self, Self::Error> {
        Ok(Self {
            tenant_id: value
                .tenant_id
                .ok_or(SerDeError::MissingField("RunAnalyzerRequest.tenant_id"))?
                .try_into()?,
            update: value
                .update
                .ok_or(SerDeError::MissingField("RunAnalyzerRequest.update"))?
                .try_into()?,
        })
    }
}

impl From<RunAnalyzerRequest> for RunAnalyzerRequestProto {
    fn from(value: RunAnalyzerRequest) -> Self {
        Self {
            tenant_id: Some(value.tenant_id.into()),
            update: Some(value.update.into()),
        }
    }
}

impl type_url::TypeUrl for RunAnalyzerRequest {
    const TYPE_URL: &'static str =
        "graplsecurity.com/graplinc.grapl.api.plugin_sdk.analyzers.v1beta1.RunAnalyzerRequest";
}

impl ProtobufSerializable for RunAnalyzerRequest {
    type ProtobufMessage = RunAnalyzerRequestProto;
}

#[derive(Debug, Clone)]
pub struct RunAnalyzerResponse {
    pub execution_result: ExecutionResult,
}

impl TryFrom<RunAnalyzerResponseProto> for RunAnalyzerResponse {
    type Error = SerDeError;
    fn try_from(value: RunAnalyzerResponseProto) -> Result<Self, Self::Error> {
        Ok(Self {
            execution_result: value
                .execution_result
                .ok_or(SerDeError::MissingField(
                    "RunAnalyzerResponse.execution_result",
                ))?
                .try_into()?,
        })
    }
}

impl From<RunAnalyzerResponse> for RunAnalyzerResponseProto {
    fn from(value: RunAnalyzerResponse) -> Self {
        Self {
            execution_result: Some(value.execution_result.into()),
        }
    }
}

impl type_url::TypeUrl for RunAnalyzerResponse {
    const TYPE_URL: &'static str =
        "graplsecurity.com/graplinc.grapl.api.plugin_sdk.analyzers.v1beta1.RunAnalyzerResponse";
}

impl ProtobufSerializable for RunAnalyzerResponse {
    type ProtobufMessage = RunAnalyzerResponseProto;
}

File: src/python/python-proto/python_proto/api/plugin_sdk/analyzers/v1beta1/messages.py
from __future__ import annotations

import dataclasses

from graplinc.grapl.api.plugin_sdk.analyzers.v1beta1 import analyzers_pb2 as proto
from python_proto import common as proto_common_msgs
from python_proto.grapl.common.v1beta1 import messages as grapl_common_msgs
from python_proto.serde import SerDe


@dataclasses.dataclass(frozen=True, slots=True)
class RunAnalyzerRequest(SerDe[proto.RunAnalyzerRequest]):
    update: Update

    _proto_cls = proto.RunAnalyzerRequest

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.RunAnalyzerRequest,
    ) -> RunAnalyzerRequest:
        return cls(
            update=Update.from_proto(proto_value.update),
        )

    def into_proto(self) -> proto.RunAnalyzerRequest:
        proto_value = self.new_proto()
        proto_value.update.CopyFrom(self.update.into_proto())
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class StringPropertyUpdate(SerDe[proto.StringPropertyUpdate]):
    uid: grapl_common_msgs.Uid
    property_name: grapl_common_msgs.PropertyName

    _proto_cls = proto.StringPropertyUpdate

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.StringPropertyUpdate,
    ) -> StringPropertyUpdate:
        return cls(
            uid=grapl_common_msgs.Uid.from_proto(proto_value.uid),
            property_name=grapl_common_msgs.PropertyName.from_proto(
                proto_value.property_name
            ),
        )

    def into_proto(self) -> proto.StringPropertyUpdate:
        proto_value = self.new_proto()
        proto_value.uid.CopyFrom(self.uid.into_proto())
        proto_value.property_name.CopyFrom(self.property_name.into_proto())
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class UInt64PropertyUpdate(SerDe[proto.UInt64PropertyUpdate]):
    uid: grapl_common_msgs.Uid
    property_name: grapl_common_msgs.PropertyName

    _proto_cls = proto.UInt64PropertyUpdate

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.UInt64PropertyUpdate,
    ) -> UInt64PropertyUpdate:
        return cls(
            uid=grapl_common_msgs.Uid.from_proto(proto_value.uid),
            property_name=grapl_common_msgs.PropertyName.from_proto(
                proto_value.property_name
            ),
        )

    def into_proto(self) -> proto.UInt64PropertyUpdate:
        proto_value = self.new_proto()
        proto_value.uid.CopyFrom(self.uid.into_proto())
        proto_value.property_name.CopyFrom(self.property_name.into_proto())
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class Int64PropertyUpdate(SerDe[proto.Int64PropertyUpdate]):
    uid: grapl_common_msgs.Uid
    property_name: grapl_common_msgs.PropertyName

    _proto_cls = proto.Int64PropertyUpdate

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.Int64PropertyUpdate,
    ) -> Int64PropertyUpdate:
        return cls(
            uid=grapl_common_msgs.Uid.from_proto(proto_value.uid),
            property_name=grapl_common_msgs.PropertyName.from_proto(
                proto_value.property_name
            ),
        )

    def into_proto(self) -> proto.Int64PropertyUpdate:
        proto_value = self.new_proto()
        proto_value.uid.CopyFrom(self.uid.into_proto())
        proto_value.property_name.CopyFrom(self.property_name.into_proto())
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class EdgeUpdate(SerDe[proto.EdgeUpdate]):
    src_uid: grapl_common_msgs.Uid
    dst_uid: grapl_common_msgs.Uid
    forward_edge_name: grapl_common_msgs.EdgeName
    reverse_edge_name: grapl_common_msgs.EdgeName

    _proto_cls = proto.EdgeUpdate

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.EdgeUpdate,
    ) -> EdgeUpdate:
        return cls(
            src_uid=grapl_common_msgs.Uid.from_proto(proto_value.src_uid),
            dst_uid=grapl_common_msgs.Uid.from_proto(proto_value.dst_uid),
            forward_edge_name=grapl_common_msgs.EdgeName.from_proto(
                proto_value.forward_edge_name
            ),
            reverse_edge_name=grapl_common_msgs.EdgeName.from_proto(
                proto_value.reverse_edge_name
            ),
        )

    def into_proto(self) -> proto.EdgeUpdate:
        proto_value = self.new_proto()
        proto_value.src_uid.CopyFrom(self.src_uid.into_proto())
        proto_value.dst_uid.CopyFrom(self.dst_uid.into_proto())
        proto_value.forward_edge_name.CopyFrom(self.forward_edge_name.into_proto())
        proto_value.reverse_edge_name.CopyFrom(self.reverse_edge_name.into_proto())
        return proto_value


UpdateInner = (
    StringPropertyUpdate | UInt64PropertyUpdate | Int64PropertyUpdate | EdgeUpdate
)


@dataclasses.dataclass(frozen=True, slots=True)
class Update(SerDe[proto.Update]):
    inner: UpdateInner

    _proto_cls = proto.Update

    @classmethod
    def from_proto(cls, proto_value: proto.Update) -> Update:
        field_name = proto_value.WhichOneof("inner")
        assert field_name is not None

        match field_name:
            case "string_property":
                return cls(
                    inner=StringPropertyUpdate.from_proto(proto_value.string_property)
                )
            case "uint64_property":
                return cls(
                    inner=UInt64PropertyUpdate.from_proto(proto_value.uint64_property)
                )
            case "int64_property":
                return cls(
                    inner=Int64PropertyUpdate.from_proto(proto_value.int64_property)
                )
            case "edge":
                return cls(inner=EdgeUpdate.from_proto(proto_value.edge))

        raise Exception(f"Unknown variant: {field_name}")

    def into_proto(self) -> proto.Update:
        msg = self.new_proto()
        match self.inner:
            case StringPropertyUpdate() as inner:
                msg.string_property.CopyFrom(inner.into_proto())
            case UInt64PropertyUpdate() as inner:
                msg.uint64_property.CopyFrom(inner.into_proto())
            case Int64PropertyUpdate() as inner:
                msg.int64_property.CopyFrom(inner.into_proto())
            case EdgeUpdate() as inner:
                msg.edge.CopyFrom(inner.into_proto())
            case _:
                raise Exception(f"Unknown variant: {self.inner}")

        return msg


@dataclasses.dataclass(frozen=True, slots=True)
class LensRef(SerDe[proto.LensRef]):
    lens_namespace: str
    lens_name: str

    _proto_cls = proto.LensRef

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.LensRef,
    ) -> LensRef:
        return cls(
            lens_namespace=proto_value.lens_namespace,
            lens_name=proto_value.lens_name,
        )

    def into_proto(self) -> proto.LensRef:
        proto_value = self.new_proto()
        proto_value.lens_namespace = self.lens_namespace
        proto_value.lens_name = self.lens_name
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class AnalyzerName(SerDe[proto.AnalyzerName]):
    value: str

    _proto_cls = proto.AnalyzerName

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.AnalyzerName,
    ) -> AnalyzerName:
        return cls(
            value=proto_value.value,
        )

    def into_proto(self) -> proto.AnalyzerName:
        proto_value = self.new_proto()
        proto_value.value = self.value
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class ExecutionHit(SerDe[proto.ExecutionHit]):
    # graph_view: GraphView
    lens_refs: list[LensRef]
    analyzer_name: AnalyzerName
    time_of_match: proto_common_msgs.Timestamp
    idempotency_key: int
    score: int

    _proto_cls = proto.ExecutionHit

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.ExecutionHit,
    ) -> ExecutionHit:
        return cls(
            lens_refs=[LensRef.from_proto(p) for p in proto_value.lens_refs],
            analyzer_name=AnalyzerName.from_proto(proto_value.analyzer_name),
            time_of_match=proto_common_msgs.Timestamp.from_proto(
                proto_value.time_of_match
            ),
            idempotency_key=proto_value.idempotency_key,
            score=proto_value.score,
        )

    def into_proto(self) -> proto.ExecutionHit:
        proto_value = self.new_proto()
        proto_value.lens_refs.extend(
            [lens_ref.into_proto() for lens_ref in self.lens_refs]
        )
        proto_value.analyzer_name.CopyFrom(self.analyzer_name.into_proto())
        proto_value.time_of_match.CopyFrom(self.time_of_match.into_proto())
        proto_value.idempotency_key = self.idempotency_key
        proto_value.score = self.score
        return proto_value


@dataclasses.dataclass(frozen=True, slots=True)
class ExecutionMiss(SerDe[proto.ExecutionMiss]):
    _proto_cls = proto.ExecutionMiss

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.ExecutionMiss,
    ) -> ExecutionMiss:
        return cls()

    def into_proto(self) -> proto.ExecutionMiss:
        proto_value = self.new_proto()
        return proto_value


ExecutionResultInner = ExecutionHit | ExecutionMiss


@dataclasses.dataclass(frozen=True, slots=True)
class ExecutionResult(SerDe[proto.ExecutionResult]):
    inner: ExecutionResultInner

    _proto_cls = proto.ExecutionResult

    @classmethod
    def from_proto(cls, proto_value: proto.ExecutionResult) -> ExecutionResult:
        field_name = proto_value.WhichOneof("inner")
        assert field_name is not None

        match field_name:
            case "hit":
                return cls(inner=ExecutionHit.from_proto(proto_value.hit))
            case "miss":
                return cls(inner=ExecutionMiss.from_proto(proto_value.miss))

        raise Exception(f"Unknown variant: {field_name}")

    def into_proto(self) -> proto.ExecutionResult:
        msg = self.new_proto()
        match self.inner:
            case ExecutionHit() as inner:
                msg.hit.CopyFrom(inner.into_proto())
            case ExecutionMiss() as inner:
                msg.miss.CopyFrom(inner.into_proto())
            case _:
                raise Exception(f"Unknown variant: {self.inner}")

        return msg


@dataclasses.dataclass(frozen=True, slots=True)
class RunAnalyzerResponse(SerDe[proto.RunAnalyzerResponse]):
    execution_result: ExecutionResult

    _proto_cls = proto.RunAnalyzerResponse

    @classmethod
    def from_proto(
        cls,
        proto_value: proto.RunAnalyzerResponse,
    ) -> RunAnalyzerResponse:
        return cls(
            execution_result=ExecutionResult.from_proto(proto_value.execution_result),
        )

    def into_proto(self) -> proto.RunAnalyzerResponse:
        proto_value = self.new_proto()
        proto_value.execution_result.CopyFrom(self.execution_result.into_proto())
        return proto_value
